File: erykandroidcommon/src/main/java/pl/edu/amu/wmi/erykandroidcommon/recycler/grouping/GroupedRecyclerFragment.java
package pl.edu.amu.wmi.erykandroidcommon.recycler.grouping;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import pl.edu.amu.wmi.erykandroidcommon.base.BaseFragment;
import pl.edu.amu.wmi.erykandroidcommon.recycler.AbstractViewHolder;


public abstract class GroupedRecyclerFragment<T extends ListItem, P extends View, U extends AbstractViewHolder<P>> extends BaseFragment {

    protected void initRecyclerView(RecyclerView recyclerView) {
        recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext()));
        GroupedRecyclerAdapter<T, P, U> groupedRecyclerAdapter = instantiateAdapter();
        recyclerView.setAdapter(groupedRecyclerAdapter);
    }

    protected abstract GroupedRecyclerAdapter<T, P, U> instantiateAdapter();

}

File: src/main/java/pl/edu/amu/wmi/erykandroidcommon/recycler/grouping/GroupedRecyclerFragment.kt
package pl.edu.amu.wmi.erykandroidcommon.recycler.grouping

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View

import pl.edu.amu.wmi.erykandroidcommon.base.BaseFragment
import pl.edu.amu.wmi.erykandroidcommon.recycler.AbstractViewHolder

abstract class GroupedRecyclerFragment<in T : ListItem, P : View, in U : AbstractViewHolder<P>> : BaseFragment() {

    protected fun initRecyclerView(recyclerView: RecyclerView) {
        recyclerView.layoutManager = LinearLayoutManager(recyclerView.context)
        val groupedRecyclerAdapter = instantiateAdapter()
        recyclerView.adapter = groupedRecyclerAdapter
    }

    protected abstract fun instantiateAdapter(): GroupedRecyclerAdapter<T, P, U>
}
